core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataDeleteIndexService.java

```diff
@@ -93,13 +105,23 @@ public ClusterState execute(final ClusterState currentState) {
                 MetaData.Builder metaDataBuilder = MetaData.builder(meta);
                 ClusterBlocks.Builder clusterBlocksBuilder = ClusterBlocks.builder().blocks(currentState.blocks());
 
+                final IndexGraveyard.Builder graveyardBuilder = IndexGraveyard.builder(metaDataBuilder.indexGraveyard());
+                final int previousGraveyardSize = graveyardBuilder.tombstones().size();
                 for (final Index index : indices) {
                     String indexName = index.getName();
                     logger.debug("[{}] deleting index", index);
                     routingTableBuilder.remove(indexName);
                     clusterBlocksBuilder.removeIndexBlocks(indexName);
                     metaDataBuilder.remove(indexName);
+                    graveyardBuilder.addTombstone(index); // add the index tombstone to the cluster state
                 }
+                final int numPurged = graveyardBuilder.purge(SETTING_MAX_TOMBSTONES.get(settings)); // purge tombstone entries if needed
```

Comment:
This is something more I meant (psuedocode):

```java
final IndexGraveyard.Builder graveyardBuilder = IndexGraveyard.builder(metaDataBuilder.indexGraveyard());
final int previousGraveyardSize = graveyardBuilder.tombstones().size();
for (final Index index : indices) {
  ... <elided> ...
  graveyardBuilder.addTombstone(index); // add the index tombstone to the cluster state
  if (previousGraveyardSize > GRAVEYARD_LIMIT) {
    graveyardBuilder.pop();
  }
}
metaDataBuilder.indexGraveyard(graveyardBuilder.build());
```

*Or*, even better, just have the `.addTombstone` method on the builder do the popping if it's over the limit, then you no longer need a `.purge` method at all, it just automatically keeps it within the configured (500) limit. I like this even more because it doesn't require the "user" of tombstones to know anything about a limit at all

This would have the nice effect of reducing the complexity of the class, as well as allowing us to use a `LinkedList` instead of doing the subList of an `ArrayList`